controller/lib/core/blower_fsm.cpp

```diff
@@ -90,92 +86,78 @@ static Duration ExpireDuration(const VentParams &params) {
 PressureControlFsm::PressureControlFsm(Time now, const VentParams &params)
     : inspire_pressure_(cmH2O(static_cast<float>(params.pip_cm_h2o))),
       expire_pressure_(cmH2O(static_cast<float>(params.peep_cm_h2o))),
-      setpoint_(expire_pressure_), start_time_(now),
-      inspire_end_(start_time_ + InspireDuration(params)),
+      start_time_(now), inspire_end_(start_time_ + InspireDuration(params)),
       expire_end_(inspire_end_ + ExpireDuration(params)) {}
 
-void PressureControlFsm::Update(Time now, const BlowerFsmInputs &inputs) {
-  if (now >= inspire_end_) {
-    inspire_finished_ = true;
-  } else {
+BlowerSystemState
+PressureControlFsm::DesiredState(Time now, const BlowerFsmInputs &inputs) {
+  if (now < inspire_end_) {
     // Go from expire_pressure_ to inspire_pressure_ over a duration of
     // RISE_TIME.  Then for the rest of the inspire time, hold at
     // inspire_pressure_.
     static_assert(RISE_TIME > milliseconds(0));
     float rise_frac = std::min(1.f, (now - start_time_) / RISE_TIME);
-    setpoint_ =
-        expire_pressure_ + (inspire_pressure_ - expire_pressure_) * rise_frac;
-  }
-  if (now >= expire_end_) {
-    finished_ = true;
-  }
-}
-
-BlowerSystemState PressureControlFsm::DesiredState() const {
-  if (inspire_finished_) {
+    return {
+        .pressure_setpoint = expire_pressure_ +
+                             (inspire_pressure_ - expire_pressure_) * rise_frac,
+        .flow_direction = FlowDirection::INSPIRATORY,
+        .pip = inspire_pressure_,
+        .peep = expire_pressure_,
+        .is_end_of_breath = false,
+    };
+  } else { // expiratory part of the cycle
     return {
         .pressure_setpoint = expire_pressure_,
         .flow_direction = FlowDirection::EXPIRATORY,
         .pip = inspire_pressure_,
         .peep = expire_pressure_,
+        .is_end_of_breath = (now >= expire_end_),
     };
   }
-  return {
-      .pressure_setpoint = setpoint_,
-      .flow_direction = FlowDirection::INSPIRATORY,
-      .pip = inspire_pressure_,
-      .peep = expire_pressure_,
-  };
 }
 
 PressureAssistFsm::PressureAssistFsm(Time now, const VentParams &params)
     : inspire_pressure_(cmH2O(static_cast<float>(params.pip_cm_h2o))),
       expire_pressure_(cmH2O(static_cast<float>(params.peep_cm_h2o))),
-      setpoint_(expire_pressure_), start_time_(now),
-      inspire_end_(start_time_ + InspireDuration(params)),
+      start_time_(now), inspire_end_(start_time_ + InspireDuration(params)),
       expire_deadline_(inspire_end_ + ExpireDuration(params)) {
   dbg_slow_flow_avg.Set(0.f);
   dbg_fast_flow_avg.Set(0.f);
 }
 
-void PressureAssistFsm::Update(Time now, const BlowerFsmInputs &inputs) {
-  if (now >= inspire_end_) {
-    inspire_finished_ = true;
-  } else {
+BlowerSystemState
+PressureAssistFsm::DesiredState(Time now, const BlowerFsmInputs &inputs) {
+  bool is_end_of_breath =
+      (now >= expire_deadline_ || PatientInspiring(now, inputs));
```

Comment:
Are you sure about this one? I kept it here because existing logic suggested that if `PatientInspiring` returns `true`, `is_end_of_breath` will be set and returned with the rest of the inspire branch state.

Reply:
But `PatientInspiring` can't be true during the inspiratory phase:

```
if (now < inspire_end_ || inputs.net_flow < ml_per_sec(0)) {
    return false;
}
```

Reply:
Fair enough